### ISSUE TITLE
WireGuard backend

### ISSUE THREAD

**Issue by kousu** [yggdrasil-network/yggdrasil-go#398]
I was thinking about cjdns and Wireguard at about the same time last month when I first appreciated what they both were. cjdns has the clever routing hack of embedding a key in an address, and Wireguard has good, fast code and strong, auditable cryptography. I would really like to have the flex of a [tincd](https://tinc-vpn.org/)-like setup but with the strength of [Wireguard](https://www.wireguard.com/).

I've been tossing around the idea of what it would take to do this. IMO it boils down to key+peer distribution: Wireguard has no way to dynamically learn new peers; but it **does** have a way to dynamically add new peers, if an outside BGP-like process runs `wg set wg0 peer key $key endpoint $ip:$port` and `ip route add ... metric ...`.

Yggdrasil's current [config file format](https://yggdrasil-network.github.io/configuration.html) covers all the same things as [Wireguard's](https://git.zx2c4.com/WireGuard/about/src/tools/man/wg-quick.8): a list of bootstrap peers plus their keys, your own key, the CIDRs that can be routed by each bootstrap peer, and the port to tunnel over. But it adds an Admin port, a firewall, two keys (#397) per node, TUN/TAP switchability, a queuing config section, and a freeform node info section. I want to suggest that the extra features you add are better supported by pre-existing -- this is the attitude WireGuard takes: just reuse the kernel's IP monitoring (`netstat`, `iftop`), the kernel's routing tables (`route add ...`), the kernel's firewall, the kernel's queue management, and only use TUN devices (which as noted in #394, is [now available for Windows](https://lists.zx2c4.com/pipermail/wireguard/2019-March/004038.html)), which should be okay because you're an IPv6 stack which is already working at Layer 3.

What WireGuard is lacking is the routing algorithm from Yggdrasil, and the mapping of public keys to node addresses.

I think there should be a way you could use wireguard as the link layer, with a side BGP-like daemon running over top of the wireguard interface that dynamically modifies the kernel routing table according to to gossip it receives from its peers.

Basically I feel like there are ways Yggdrasil and Wireguard can learn from each other.

**Comment by neilalexander** [comment]
Yggdrasil basically reimplements both the switch layer and the router layer, so traffic moving through the network is switched according to the spanning tree and traffic to/from the current node passes through the router up to userspace via TUN/TAP.

One of the beauties of this design is that the routing layer is never made aware of traffic flowing via your node en route from somewhere else to somewhere else. This just isn't needed - the switch takes care of forwarding those packets towards their destination, so the router is left alone in that case.

We've talked a lot in the past about how having the switch element (at a minimum) implemented in kernel space could significantly increase our performance. I suspect this is will be a future goal of the project eventually.

The approach of dynamically configuring Wireguard and the kernel routing table wouldn't work for us, because it breaks some core principles in our design:

1. Traffic not destined for the node would need to escape the switch in order to be compared against the kernel routing table, which won't make sense because...
1. The Yggdrasil switch isn't routed by next-hop IPs and instead forwards towards spanning tree coordinates, which the kernel routing table has no understanding of (because it's not IP)
1. The router layer then disappears somewhere and we need to come up with some different solution for establishing and encrypting session traffic (but where?)

As I see it Yggdrasil will pretty much only work if it can be implemented as a stack.

**Comment by kousu** [comment]
Hey, I'd just like to thank you again for answering so completely and letting some stranger on the internet pick your brain :). Some of this context would improve the whitepaper.

I'm going to read this closer tonight and give it a think.

**Comment by neilalexander** [comment]
Closing this now just to clean up open issues.
